janus_server/src/aggregator/accumulator.rs

```diff
@@ -0,0 +1,165 @@
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub(super) struct Accumulator<A: vdaf::Aggregator>
+where
+    for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+{
+    task_id: TaskId,
+    min_batch_duration: Duration,
+    #[derivative(Debug = "ignore")]
+    aggregation_param: A::AggregationParam,
+    accumulations: HashMap<Time, Accumulate<A>>,
+}
+
+impl<A: vdaf::Aggregator> Accumulator<A>
+where
+    for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+{
+    /// Create a new accumulator
+    pub(super) fn new(
+        task_id: TaskId,
+        min_batch_duration: Duration,
+        aggregation_param: &A::AggregationParam,
+    ) -> Self {
+        Self {
+            task_id,
+            min_batch_duration,
+            aggregation_param: aggregation_param.clone(),
+            accumulations: HashMap::new(),
+        }
+    }
+
+    /// Update the in-memory accumulators with the provided output share and report nonce.
+    pub(super) fn update(
+        &mut self,
+        output_share: &A::OutputShare,
+        nonce: Nonce,
+    ) -> Result<(), datastore::Error> {
+        let key = nonce
+            .time()
+            .to_batch_unit_interval_start(self.min_batch_duration)
+            .map_err(|e| datastore::Error::User(e.into()))?;
+        if let Some(accumulate) = self.accumulations.get_mut(&key) {
+            accumulate
+                .update(output_share, nonce)
+                .map_err(|e| datastore::Error::User(e.into()))?;
+        } else {
+            self.accumulations.insert(
+                key,
+                Accumulate {
+                    aggregate_share: A::AggregateShare::from(output_share.clone()),
+                    report_count: 1,
+                    checksum: NonceChecksum::from_nonce(nonce),
+                },
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Write the accumulated aggregate shares, report counts and checksums to the datastore. If a
+    /// batch unit aggregation already exists for some accumulator, it is updated. If no batch unit
+    /// aggregation exists, one is created and initialized with the accumulated values.
+    #[tracing::instrument(skip(self, tx), err)]
+    pub(super) async fn flush_to_datastore<E>(
+        self,
+        tx: &Transaction<'_>,
+    ) -> Result<(), datastore::Error>
+    where
+        for<'a> A::AggregateShare: TryFrom<&'a [u8], Error = E>,
+        E: std::fmt::Display,
+    {
+        for (unit_interval_start, accumulate) in self.accumulations {
+            let unit_interval = Interval::new(unit_interval_start, self.min_batch_duration)?;
+
+            let mut batch_unit_aggregations = tx
+                .get_batch_unit_aggregations_for_task_in_interval::<A, _>(
+                    self.task_id,
+                    unit_interval,
+                    &self.aggregation_param,
+                )
+                .await?;
+
+            if batch_unit_aggregations.len() > 1 {
```

Comment:
nit, optional: we could make a separate tx getter that, instead of finding all batch unit aggregations over a given interval, instead takes a point in time and returns the batch unit interval (if any) including that point in time].

doing so would remove this error case, and result in slightly simpler queries.

Reply:
I see your point, but I prefer adding a ~6 line `if` statement here over adding a large method to `Transaction`.